README.md
# responsive-typescale

> Responsive typography made easy. Different modular scales at each breakpoint.

responsive-typescale outputs responsive CSS as JavaScript strings. It is designed to be used with modern CSS-in-JS libraries such as [styped-components](https://github.com/styled-components/styled-components) or [emotion](https://github.com/emotion-js/emotion). Alternatively, the outputted strings can be appended onto the `<head>` directly, for example using [insert-css](https://github.com/substack/insert-css).

## Install

```
$ npm install responsive-typescale
```

## Usage

```js
import initTypescale from 'responsive-typescale';

const breakpoints = {
    default: {
        base: 1.15,
        ratio: 1.2,
        lineHeight: 1.4
    },
    extraLarge: {
        width: 1440,
        base: 1.15,
        ratio: 1.175,
        lineHeight: 1.4
    },
    large: {
        width: 1080,
        base: 1.125,
        ratio: 1.175,
        lineHeight: 1.35
    },
    medium: {
        width: 720,
        base: 1.1,
        ratio: 1.15,
        lineHeight: 1.3
    },
    small: {
        width: 480,
        base: 1,
        ratio: 1.1,
        lineHeight: 1.3
    }
};

const typescale = initTypescale(breakpoints);
typescale.size(7);
```

String output:

```css
font-size: 4.121rem;
@media (max-width: 90em) {
    font-size: 3.556rem;
}
@media (max-width: 67.5em) {
    font-size: 3.479rem;
}
@media (max-width: 45em) {
    font-size: 2.926rem;
}
@media (max-width: 30em) {
    font-size: 1.949rem;
}
```

## API

### initTypescale(breakpoints?)

Creates a `responsive-typescale` instance that uses the passed breakpoints. Returns an object with the following functions:

* [size](#inittypescalesizescalelevel)
* [padding](#inittypescalepaddingdirection-rhythmunits)
* [margin](#inittypescalemargindirection-rhythmunits)
* [media](#inittypescalemediabreakpointname-css)

#### breakpoints

Type: `object` - (TS: [`Breakpoints`](src/lib/breakpoints.ts#L7))

Default: [`sensibleDefaultBreakpoints`](src/lib/breakpoints.ts#L12)

An object containing each [`Breakpoint`](src/lib/breakpoints.ts#L1). A `Breakpoint` contains a `width` key indicating when it activates, and its associated modular scale. The names of the breakpoints are used in the [media](#inittypescalemediabreakpointname-css) function.

Note: It is required to have a Breakpoint with the key `default`. The `default` breakpoint does not need to have a `width` key.

### initTypescale().size(scaleLevel, rhythmUnits?)

Returns a `string` with proper CSS that sets the `font-size` to the passed modular scale level. Uses the passed breakpoint from `initTypescale`, and the returned string contains the proper media queries for each breakpoint. Optionally, when `rhythmUnits` argument is provided, it also sets the `line-height` to the specified rhythm unit amount.

#### scaleLevel

Type: `number`

The level of the modular scale, starting from `0`, which is the base.

#### rhythmUnits

Type: `number`

This argument will add `line-height` to the returned CSS string. A single rhythm unit is equal to the line-height of the body text (modular scale level of 0).

### initTypescale().padding(direction, rhythmUnits)

Returns a `string` with proper CSS that sets `padding` in the specified direction by the specified rhythm amount.  Uses the passed breakpoint from `initTypescale`, and the returned string contains the proper media queries for each breakpoint.

#### direction

Type: `string` - (TS: [`Direction`](src/lib/spacing.ts#L4))

Direction to apply the padding: `top`, `bottom`, `left`, or `right`.

#### rhythmUnits

Type: `number`

A single rhythm unit is equal to the line-height of the body text (modular scale level of 0).

### initTypescale().margin(direction, rhythmUnits)

Same as the [`padding`](#inittypescalepaddingdirection-rhythmunits) function, except it sets the `margin` CSS property.

### initTypescale().media[breakpointName] (css)

Helper function to generate CSS strings that target the specified breakpoint.

#### breakpointName

Type: `string`

Name (object key) of the breakpoint.

#### css

Type: `string`

Example (using a `responsive-typescale` instance that was initialized with the [`sensibleDefaultBreakpoints`](src/lib/breakpoints.ts#L12)):

```js
media.large('color: red;')
```

String output:

```css
@media (max-width: 67.5em) {
    color: red;
}
```

## Tips

### Exporting the `responsive-typescale` Instance

It's a good idea to initialize the library using initTypescale, and then export the instance.

`typescale.js`

```js
import initTypescale from 'responsive-typescale';

const typescale = initTypescale();

export const { size, margin, media, padding } = typescale;
```

`anotherFile.js`

```js
import { size, padding } from './typescale.js';

const componentStyle = `
    ${size(0, 1)}
    ${padding('top', 3)}
`;
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.1](https://github.com/KennethWangDotDev/responsive-typescale/compare/v2.0.0...v2.0.1) (2019-08-18)


### Bug Fixes

* **lib/spacing.ts:** fixed incorrect spacing outputs ([98c7aa0](https://github.com/KennethWangDotDev/responsive-typescale/commit/98c7aa0))



## [2.0.0](https://github.com/KennethWangDotDev/responsive-typescale/compare/v1.1.5...v2.0.0) (2019-08-17)


### Features

* use any number for modular scale size, and extra arg for line-height ([1ea4084](https://github.com/KennethWangDotDev/responsive-typescale/commit/1ea4084))


### BREAKING CHANGES

* initTypescale().size() no longer accepts semantic aliases like `p`



### [1.1.5](https://github.com/KennethWangDotDev/responsive-typescale/compare/v1.1.4...v1.1.5) (2019-08-14)



### [1.1.4](https://github.com/KennethWangDotDev/responsive-typescale/compare/v1.1.3...v1.1.4) (2019-08-14)



### [1.1.3](https://github.com/KennethWangDotDev/responsive-typescale/compare/v1.1.2...v1.1.3) (2019-08-11)



### [1.1.2](https://github.com/KennethWangDotDev/responsive-typescale/compare/v1.1.1...v1.1.2) (2019-08-11)



### [1.1.1](https://github.com/KennethWangDotDev/responsive-typescale/compare/v1.1.0...v1.1.1) (2019-08-11)


### Bug Fixes

* **package.json:** remove unused typedoc publishing feature ([c910da6](https://github.com/KennethWangDotDev/responsive-typescale/commit/c910da6))



## 1.1.0 (2019-08-11)


### Features

* **responsive-typescale:** initial commit ([45737f1](https://github.com/KennethWangDotDev/responsive-typescale/commit/45737f1))


src/tests/size.spec.ts
import test from 'ava';
import dedent from 'dedent';
import { testBreakpoint } from './test-breakpoint';
import { generateTypescale } from '../lib/typescale';
import { generateMediaBreakpoints } from '../lib/media';
import { generateSizeMixin } from '../lib/size';

test('generates size mixin correctly', t => {
    const typescale = generateTypescale(testBreakpoint);
    const media = generateMediaBreakpoints(testBreakpoint);
    const size = generateSizeMixin(testBreakpoint, typescale, media);
    t.is(
        size(0),
        dedent`
        font-size: 1.15rem;
        @media (max-width: 90em) {
            font-size: 1.15rem;
        }
        @media (max-width: 67.5em) {
            font-size: 1.125rem;
        }
        @media (max-width: 45em) {
            font-size: 1.1rem;
        }
        @media (max-width: 30em) {
            font-size: 1rem;
        }\n
    `
    );
    t.is(
        size(7, 2),
        dedent`
        font-size: 4.121rem;
        line-height: 3.2200rem;
        @media (max-width: 90em) {
            font-size: 3.556rem;
        line-height: 3.2200rem;
        }
        @media (max-width: 67.5em) {
            font-size: 3.479rem;
        line-height: 3.0375rem;
        }
        @media (max-width: 45em) {
            font-size: 2.926rem;
        line-height: 2.8600rem;
        }
        @media (max-width: 30em) {
            font-size: 1.949rem;
        line-height: 2.6000rem;
        }\n
    `
    );
});


src/tests/media.spec.ts
import test from 'ava';
import dedent from 'dedent';
import { generateMediaBreakpoints } from '../lib/media';
import { testBreakpoint } from './test-breakpoint';

test('generates media queries correctly', t => {
    const media = generateMediaBreakpoints(testBreakpoint);
    t.is(
        media.large('color: red;'),
        dedent`
        @media (max-width: 67.5em) {
            color: red;
        }
    `
    );
    t.is(
        media.medium('color: blue;'),
        dedent`
        @media (max-width: 45em) {
            color: blue;
        }
    `
    );
});


src/tests/spacing.spec.ts
import test from 'ava';
import dedent from 'dedent';
import { testBreakpoint } from './test-breakpoint';
import { generateMediaBreakpoints } from '../lib/media';
import { generateSpacingMixins } from '../lib/spacing';

test('generates spacing mixins correctly', t => {
    const media = generateMediaBreakpoints(testBreakpoint);
    const [margin, padding] = generateSpacingMixins(testBreakpoint, media);
    t.is(
        margin('bottom', 1),
        dedent`
        margin-bottom: 1.6100rem;
        @media (max-width: 90em) {
            margin-bottom: 1.6100rem;
        }
        @media (max-width: 67.5em) {
            margin-bottom: 1.5188rem;
        }
        @media (max-width: 45em) {
            margin-bottom: 1.4300rem;
        }
        @media (max-width: 30em) {
            margin-bottom: 1.3000rem;
        }\n
    `
    );
    t.is(
        padding('top', 4),
        dedent`
        padding-top: 6.4400rem;
        @media (max-width: 90em) {
            padding-top: 6.4400rem;
        }
        @media (max-width: 67.5em) {
            padding-top: 6.0750rem;
        }
        @media (max-width: 45em) {
            padding-top: 5.7200rem;
        }
        @media (max-width: 30em) {
            padding-top: 5.2000rem;
        }\n
    `
    );
});


src/lib/size.ts
import { MediaBreakpoints } from './media';
import { TypeScaleBreakpoints } from './typescale';
import { Breakpoints } from './breakpoints';

const sizeFunction = (
    scaleLevel: number,
    rhythmUnits: number,
    breakpoints: Breakpoints,
    typescaleBreakpoints: TypeScaleBreakpoints,
    mediaBreakpoints: MediaBreakpoints
): string => {
    let sizeCss = '';
    Object.entries(typescaleBreakpoints).forEach(([breakpointKey, typescale]) => {
        // Generate CSS Rule
        let rule = `${`font-size: ${typescale(scaleLevel)}rem;`}\n`;
        if (rhythmUnits > 0) {
            rule += `line-height: ${(
                breakpoints[breakpointKey].lineHeight *
                breakpoints[breakpointKey].base *
                rhythmUnits
            ).toFixed(4)}rem;\n`;
        }
        // Generate responsive CSS
        if (breakpointKey === 'default') {
            sizeCss += rule;
        } else {
            sizeCss += `${mediaBreakpoints[breakpointKey](rule)}\n`;
        }
    });
    return sizeCss;
};

const generateSizeMixin = (
    breakpoints: Breakpoints,
    typescaleBreakpoints: TypeScaleBreakpoints,
    mediaBreakpoints: MediaBreakpoints
) => {
    return (scaleLevel: number, rhythmUnits: number = 0) =>
        sizeFunction(scaleLevel, rhythmUnits, breakpoints, typescaleBreakpoints, mediaBreakpoints);
};

export { generateSizeMixin };


src/lib/spacing.ts
import { MediaBreakpoints } from './media';
import { Breakpoints } from './breakpoints';

type Direction = 'top' | 'bottom' | 'left' | 'right';
type SpacingType = 'margin' | 'padding';

const spacingFunction = (
    spacingType: SpacingType,
    direction: Direction,
    rhythmUnits: number,
    breakpoint: Breakpoints,
    mediaBreakpoints: MediaBreakpoints
) => {
    let spacingCss = '';
    Object.entries(breakpoint).forEach(([breakpointKey, breakpointValue]) => {
        const rule = `${spacingType}-${direction}: ${(
            breakpointValue.base *
            breakpointValue.lineHeight *
            rhythmUnits
        ).toFixed(4)}rem;\n`;
        if (breakpointKey === 'default') {
            spacingCss += rule;
        } else {
            spacingCss += `${mediaBreakpoints[breakpointKey](rule)}\n`;
        }
    });
    return spacingCss;
};

const generateSpacingMixins = (breakpoint: Breakpoints, mediaBreakpoints: MediaBreakpoints) => {
    const margin = (direction: Direction, rhythmUnits: number) => {
        return spacingFunction('margin', direction, rhythmUnits, breakpoint, mediaBreakpoints);
    };
    const padding = (direction: Direction, rhythmUnits: number) => {
        return spacingFunction('padding', direction, rhythmUnits, breakpoint, mediaBreakpoints);
    };
    return [margin, padding];
};

export { generateSpacingMixins };


src/lib/breakpoints.ts
interface Breakpoint {
    readonly width?: number;
    readonly base: number;
    readonly ratio: number;
    readonly lineHeight: number;
}
interface Breakpoints {
    readonly default: Breakpoint;
    readonly [breakpointName: string]: Breakpoint;
}

const sensibleDefaultBreakpoints: Breakpoints = {
    default: {
        base: 1.15,
        ratio: 1.2,
        lineHeight: 1.4
    },
    extraLarge: {
        width: 1440,
        base: 1.15,
        ratio: 1.175,
        lineHeight: 1.4
    },
    large: {
        width: 1080,
        base: 1.125,
        ratio: 1.175,
        lineHeight: 1.35
    },
    medium: {
        width: 720,
        base: 1.1,
        ratio: 1.15,
        lineHeight: 1.3
    },
    small: {
        width: 480,
        base: 1,
        ratio: 1.1,
        lineHeight: 1.3
    }
};

export { sensibleDefaultBreakpoints, Breakpoint, Breakpoints };
